### ISSUE TITLE
do you need to put decoder.dropout at testing time?

### ISSUE THREAD

**Issue by homelifes** [sgrvinod/a-PyTorch-Tutorial-to-Image-Captioning#71]
Hi. I've realized that you've removed the line of code: decoder.dropout(h) at testing time. At inference, the dropout behaves differently (gets scaled to account for the loss in training). However, you are not informing the network of any dropout at testing time, so how should it supposed to be scaled? Because when you are removing the decoder.dropout, you are getting the unscaled values. They yield different results. Have a look at the code below:

```
import torch
import torch.nn as nn

x = torch.randn(4,4)
drop = nn.Dropout(0.5)
model = nn.Linear(4,6)
model.train()
model(drop(x))

model.eval()
d = model(drop(x))
n = model(x)
print(d == n)
```
This gives a matrix of 0s (all False).

![Capture](https://user-images.githubusercontent.com/46392773/56857449-5435b300-69a0-11e9-8840-5788f9a2941a.PNG)

Thanks!

**Comment by homelifes** [comment]
@sgrvinod may you please answer?

**Comment by sgrvinod** [comment]
Hi, sorry, I haven't had to chance to reply to many issues.

As I understand it, what you're asking is, when Dropout is applied during training with a probability `p` to keep a value, then we must scale the values by a factor of `p` during inference/testing because `1 - p` of the values are no longer being dropped out. So, since I'm not using the Dropout layer during inference, the network doesn't know to scale values by `p`. Am I right in understanding your question?

In PyTorch, however, the Dropout layer is structured such that the scaling is applied _during training_ and not during inference. So the not-dropped values during training are scaled by a factor of `1 / p`, instead of scaling all the values during inference by a factor of `p`. This is convenient because the scaling is applied when Dropout is actually applied (i.e. during training), and not when Dropout is not applied (i.e. during inference), and therefore we don't need to worry about it at test time.  

In your code, the Dropout layer `drop` is not actually a part of `model`, so regardless of whether `model` is in `train()` or `eval()` mode, `drop` is still in `train()` mode. So, what `drop` is doing to `x` is that it is dropping half the values (`p = 0.5`), and multiplying the not-dropped values by `1 / 0.5`, i.e. doubling them. So `drop(x)` and `x` have no equal elements, which is why `model(drop(x))` and `model(x)` have no equal elements.

You can test the behaviour of the PyTorch Dropout in `train()` and `eval()` modes:

```python
drop = nn.Dropout(0.5)
x = torch.randn(1, 5)
print(x)

drop.train()  # unnecessary because it's already in train mode, since it's the default
dropped_train = drop(x)
print(dropped_train)  # dropout in train() scales the not-dropped values by 2 (= 1 / 0.5)
print(dropped_train == 2 * x)  # 1s where values weren't dropped

drop.eval()
dropped_eval = drop(x)
print(dropped_eval)  # dropout in eval() mode has no effect
print(dropped_eval == x)  # all 1s
```
Similarly,

```python
x = torch.randn(1, 5)
model = nn.Sequential(*[nn.Linear(5, 5), nn.Dropout(0.5)])

model.train()
train_result = model(x)

model.eval()
eval_result = model(x)

print(train_result == 2 * eval_result)  # 1s where values weren't dropped
```
Hope this answers your question. If I misunderstood, let me know.

**Comment by homelifes** [comment]
Thank you for your great answer! I completetly forgot that PyTorch uses inverted dropout (which leaves the values of testing time unchanged). Thanks a lot for the examples as well!
